diff --git a/py_scripts/GenerateAll.py b/py_scripts/GenerateAll.py
index cd01aa2..de0eaf2 100644
--- a/py_scripts/GenerateAll.py
+++ b/py_scripts/GenerateAll.py
@@ -21,6 +21,7 @@ try:
 	runPartial = False
 	isPalladium = False
 	TypeOfKey = "openssl"
+	useSignedCombo0 = None
 
 
 	index = 1
@@ -67,6 +68,7 @@ try:
 		#option 1) .\py_scripts\GenerateAll.py openssl\HSM  0 partial
 		#option 2) .\py_scripts\GenerateAll.py openssl\HSM  0 palladium
 		#option 3) .\py_scripts\GenerateAll.py openssl\HSM  0 partial palladium
+		#option 4) .\py_scripts\GenerateAll.py openssl\HSM  [pre-signed Combo0 path]
 
 
 	if (runPartial):
@@ -78,7 +80,9 @@ try:
 		print("==========================================================")
 		print("== Generate All Images")
 		print("==========================================================")
-		ImageGeneration.GenerateImages.Run(TypeOfKey, pinCode, isPalladium)
+		if component_num != "0":
+			useSignedCombo0 = component_num
+		ImageGeneration.GenerateImages.Run(TypeOfKey, pinCode, isPalladium, useSignedCombo0)
 
 except Exception as e:
 	exc_type, exc_obj, exc_tb = sys.exc_info()
diff --git a/py_scripts/ImageGeneration/GenerateImages.py b/py_scripts/ImageGeneration/GenerateImages.py
index 1937e21..a6a18dc 100644
--- a/py_scripts/ImageGeneration/GenerateImages.py
+++ b/py_scripts/ImageGeneration/GenerateImages.py
@@ -24,7 +24,7 @@ from .IGPS_common import *
 from .Register_csv_parse import *
 
 
-def Run(TypeOfKey, pinCode,isPalladium):
+def Run(TypeOfKey, pinCode, isPalladium, useSignedCombo0):
 	currpath = os.getcwd()
 	os.chdir(os.path.dirname(os.path.abspath(__file__)))
 	
@@ -76,7 +76,7 @@ def Run(TypeOfKey, pinCode,isPalladium):
 			Sign_combo0(TypeOfKey, pinCode, isPalladium, TypeOfKey_TIP, TypeOfKey_BMC)
 			Sign_combo1(TypeOfKey, pinCode, isPalladium, TypeOfKey_TIP, TypeOfKey_BMC)
 			
-		Merge_signed_files(isPalladium)
+		Merge_signed_files(isPalladium, useSignedCombo0)
 
 	except (Exception) as e:
 		exc_type, exc_obj, exc_tb = sys.exc_info()
diff --git a/py_scripts/ImageGeneration/IGPS_common.py b/py_scripts/ImageGeneration/IGPS_common.py
index fb5b4ed..f902446 100755
--- a/py_scripts/ImageGeneration/IGPS_common.py
+++ b/py_scripts/ImageGeneration/IGPS_common.py
@@ -139,15 +139,28 @@ def Uboot_header_embed_pointers_to_all_fw():
 		# Replace_binary_array(UbootAndHeader_bin, 0x1BC, [0xAA, 0xBB, 0xCC, 0xDD], 4, False)
 
 
-def MergeBinFilesAndPadAndPrint(isPalladium):
+def MergeBinFilesAndPadAndPrint(isPalladium, useSignedCombo0: str = None):
 	# Merge files
 	bbS = COMBO1_OFFSET
-	
-	tipS_L0 =      Merge_bin_files_and_pad(KmtAndHeader_bin            , TipFwAndHeader_L0_bin    , Kmt_TipFwL0_bin                       , 0x1000, 0x20)
-	sa_tipS_L0 =   Merge_bin_files_and_pad(KmtAndHeader_bin            , SA_TipFwAndHeader_L0_bin , SA_Kmt_TipFwL0_bin                    , 0x1000, 0x20)
-	tipS_L0_UT =   Merge_bin_files_and_pad(KmtAndHeader_bin            , TipFwAndHeader_L0_UT_bin , Kmt_TipFwL0_UT_bin                    , 0x1000, 0x20)
-	skmtS  =       Merge_bin_files_and_pad(Kmt_TipFwL0_bin             , SkmtAndHeader_bin        , Kmt_TipFwL0_Skmt_bin                  , 0x1000, 0x20)
-	tipS_L1 =      Merge_bin_files_and_pad(Kmt_TipFwL0_Skmt_bin        , TipFwAndHeader_L1_bin    , Kmt_TipFwL0_Skmt_TipFwL1_bin          , 0x1000, 0x20)
+	if useSignedCombo0 is None:
+		tipS_L0 =      Merge_bin_files_and_pad(KmtAndHeader_bin            , TipFwAndHeader_L0_bin    , Kmt_TipFwL0_bin                       , 0x1000, 0x20)
+		sa_tipS_L0 =   Merge_bin_files_and_pad(KmtAndHeader_bin            , SA_TipFwAndHeader_L0_bin , SA_Kmt_TipFwL0_bin                    , 0x1000, 0x20)
+		tipS_L0_UT =   Merge_bin_files_and_pad(KmtAndHeader_bin            , TipFwAndHeader_L0_UT_bin , Kmt_TipFwL0_UT_bin                    , 0x1000, 0x20)
+		skmtS  =       Merge_bin_files_and_pad(Kmt_TipFwL0_bin             , SkmtAndHeader_bin        , Kmt_TipFwL0_Skmt_bin                  , 0x1000, 0x20)
+		tipS_L1 =      Merge_bin_files_and_pad(Kmt_TipFwL0_Skmt_bin        , TipFwAndHeader_L1_bin    , Kmt_TipFwL0_Skmt_TipFwL1_bin          , 0x1000, 0x20)
+	else:
+		# TODO: we don't have offset for TIP L0, L1, and skmt. Do we need to calculate them?
+		try:
+			print(f"Copying signed combo0 {useSignedCombo0} to {Kmt_TipFwL0_Skmt_TipFwL1_bin}")
+			copyfile(useSignedCombo0, Kmt_TipFwL0_Skmt_TipFwL1_bin)
+		except Exception as e:
+			print(f"current path: {os.getcwd()}")
+			print("Failed to copy file: ", e)
+			raise
+		tipS_L0 = 0
+		skmtS = 0
+		tipS_L1 = 0
+
 	bbS =          Merge_bin_files_and_pad(Kmt_TipFwL0_Skmt_TipFwL1_bin, BootBlockAndHeader_bin   , Kmt_TipFwL0_Skmt_TipFwL1_BootBlock_bin, bbS   , 0x20)
 	
 	# check that bootblock is still at 512KB offset
@@ -173,7 +186,8 @@ def MergeBinFilesAndPadAndPrint(isPalladium):
 	
 	# os.remove(tmp_bin)
 	# os.remove(Kmt_TipFwL0_bin)
-	os.remove(Kmt_TipFwL0_Skmt_bin)
+	if useSignedCombo0 is None:
+		os.remove(Kmt_TipFwL0_Skmt_bin)
 	os.remove(BootBlock_BL31_bin)
 	os.remove(BootBlock_BL31_OpTee_bin)
 	os.remove(Kmt_TipFwL0_Skmt_TipFwL1_BootBlock_bin)
@@ -560,13 +574,13 @@ def Sign_combo1(TypeOfKey, pinCode, isPalladium, TypeOfKey_TIP=None, TypeOfKey_B
 
 
 
-def Merge_signed_files(isPalladium):
+def Merge_signed_files(isPalladium, useSignedCombo0):
 
 	currpath = os.getcwd()
 	os.chdir(os.path.dirname(os.path.abspath(__file__)))
 
 	try:
-		MergeBinFilesAndPadAndPrint(isPalladium)
+		MergeBinFilesAndPadAndPrint(isPalladium, useSignedCombo0)
 		# Move Secure images to Secure Directory
 		MoveToFolder(isPalladium, secure_outputs_dir)
 		# need to remove the no tip secure bin that was created in last operation of moving all from basic to secure
