From a2d68242c1701165f64bac0a718da976d5a4d0da Mon Sep 17 00:00:00 2001
From: Joseph Liu <kwliu@nuvoton.com>
Date: Mon, 7 Mar 2022 16:26:54 +0800
Subject: [PATCH] kernel: m1120 dts

Signed-off-by: Joseph Liu <kwliu@nuvoton.com>
---
 arch/arm64/boot/dts/nuvoton/Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/arm64/boot/dts/nuvoton/Makefile b/arch/arm64/boot/dts/nuvoton/Makefile
index a99dab90472a..f99dff38f708 100644
--- a/arch/arm64/boot/dts/nuvoton/Makefile
+++ b/arch/arm64/boot/dts/nuvoton/Makefile
@@ -1,2 +1,3 @@
 # SPDX-License-Identifier: GPL-2.0
 dtb-$(CONFIG_ARCH_NPCM) += nuvoton-npcm845-evb.dtb
+dtb-$(CONFIG_ARCH_NPCM) += m1120-c219b.dtb
-- 
2.34.1

